compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -136,6 +237,128 @@ emitLinkedTuningSpec(ModuleOp module, ArrayRef<NamedSequenceOp> specsToLink) {
+        innerModule, namedSequenceOpsToMove, foreachMatchOps,
+        namedSequenceToForeachMatch);
+  }
+
+  // Step 2-a: Make sure the name sequence names are unique, and then move
+  // collected NamedSequenceOps to the top-level module.
+  Type anyOpType = builder.getType<transform::AnyOpType>();
+  llvm::StringMap<unsigned> specNameCounts;
+  for (NamedSequenceOp op : namedSequenceOpsToMove) {
+    StringRef specName = op.getSymName();
+    std::string newSpecName = getUniqueSpecName(specName, specNameCounts);
+    op.setSymName(newSpecName);
+
+    // Skip updating ForeachMatchOp if the NamedSequenceOp is not used in it
+    // or its name has not changed.
+    if (!namedSequenceToForeachMatch.contains(op) || newSpecName == specName) {
+      op.getOperation()->moveBefore(module.getBody(), module.getBody()->end());
+      continue;
+    }
+
+    ForeachMatchOp foreachMatchOp = namedSequenceToForeachMatch[op];
+
+    SmallVector<Attribute> updatedMatchers, updatedActions;
+    for (auto matcherAttr : foreachMatchOp.getMatchers()) {
+      StringRef matcherName =
+          cast<SymbolRefAttr>(matcherAttr).getRootReference();
+      updatedMatchers.push_back(
+          (matcherName == specName)
+              ? SymbolRefAttr::get(builder.getContext(), newSpecName)
+              : matcherAttr);
+    }
+
+    for (auto actionAttr : foreachMatchOp.getActions()) {
+      StringRef actionName = cast<SymbolRefAttr>(actionAttr).getRootReference();
+      updatedActions.push_back(
+          (actionName == specName)
+              ? SymbolRefAttr::get(builder.getContext(), newSpecName)
+              : actionAttr);
+    }
+
+    // Apply the updated matchers and actions.
+    foreachMatchOp.setMatchersAttr(builder.getArrayAttr(updatedMatchers));
+    foreachMatchOp.setActionsAttr(builder.getArrayAttr(updatedActions));
+
+    op.getOperation()->moveBefore(module.getBody(), module.getBody()->end());
+  }
+  // Step 2-b: Create a new NamedSequenceOp `__kernel_config` in the top-level
+  // module.
+  builder.setInsertionPointToEnd(module.getBody());
+  Location loc = module.getLoc();
+  auto newNamedSequence =
+      createKernelConfigOp(builder, loc, kKernelConfigSpecName);
+
+  bool hasConsumedArg = llvm::any_of(foreachMatchOps, hasConsumedArgument);
+
+  StringRef attrName =
+      hasConsumedArg ? kArgConsumedAttrName : kArgReadOnlyAttrName;
+  newNamedSequence.setArgAttr(0, attrName, builder.getUnitAttr());
+  newNamedSequence->setAttr(kTuningSpecEntrypointAttrName,
+                            builder.getUnitAttr());
+  // Indicate that the output module is a default tuning spec after merging.
+  module->setAttr(kTuningSpecDefaultEntrypointAttrName, builder.getUnitAttr());
+
+  // Step 2-C: Create a new block inside the NamedSequenceOp and merge the
+  // ForeachMatchOp from each inner module into one ForachMatchOp.
+  SmallVector<Type, 4> resultTypes;
+  llvm::append_range(resultTypes, TypeRange{anyOpType});
```

Comment:
Do we need the `TypeRange`?
```suggestion
  SmallVector<Type, 4> resultTypes = {anyOpType};
```

This should work, no?